### PR TITLE
SPISlave.end() added

### PR DESCRIPTION
**Pull request by jandrassy**
resetting the registers is needed for the switch to master mode for example for AVRISP

### CODE REVIEW

File: libraries/SPISlave/src/SPISlave.cpp [hunk 1]
```diff
@@ -72,6 +72,14 @@ void SPISlaveClass::begin()
     hspi_slave_onStatusSent(&_s_status_tx);
     hspi_slave_begin(4, this);
 }
+void SPISlaveClass::end()
+{
+    hspi_slave_onData(NULL);
```
Comment (devyte) [inline libraries/SPISlave/src/SPISlave.cpp]: Please use nullptr instead of NULL

File: libraries/SPISlave/src/hspi_slave.c [hunk 2]
```diff
@@ -98,6 +98,23 @@ void hspi_slave_begin(uint8_t status_len, void * arg)
     ETS_SPI_INTR_ENABLE();
 }
 
+void hspi_slave_end()
+{
+  ETS_SPI_INTR_DISABLE();
+  ETS_SPI_INTR_ATTACH(_hspi_slave_isr_handler, NULL);
```
Comment (devyte) [inline libraries/SPISlave/src/hspi_slave.c]: Shouldn't this be something like:
`ETS_SPI_INTR_ATTACH(nullptr, nullptr);`
?

Comment (jandrassy) [reply libraries/SPISlave/src/hspi_slave.c]: sorry, the first argument should be NULL. I change it.

it is a C source code and nullptr is not declared

File: libraries/SPISlave/src/SPISlave.cpp (hunk 1 above)
Comment (jandrassy) [reply libraries/SPISlave/src/SPISlave.cpp]: the hspi_slave_* functions are C functions called from C++. should nullptr  be used?

File: libraries/SPISlave/src/hspi_slave.c (hunk 2 above)
Comment (devyte) [reply libraries/SPISlave/src/hspi_slave.c]: Ah yes, C file, so NULL it is.

File: libraries/SPISlave/src/SPISlave.cpp (hunk 1 above)
Comment (devyte) [reply libraries/SPISlave/src/SPISlave.cpp]: Yes. NULL is a #define 0 in C, so an int that gets cast into a 32bit ptr, and in C++ it is a #define 0x00 that is cast into a void*, while nullptr is a C++ builtin that is really a null ptr. It's a type and/or style thing. As a rule of thumb, use of nullptr is considered better in C++ code. You can google nullptr vs. NULL to read about the C++ decision.

**Comment by jandrassy** [comment]
I made the requested changes as a new commit. How can we do it to look like one commit?

**Comment by jandrassy** [comment]
I made the changes and changed the commit.

**Comment by svofski** [comment]
I very much would love having this patch merged. Thanks!

**Comment by jandrassy** [comment]
@svofski try to add your request as issue, with a reason why you need this

**Comment by jandrassy** [comment]
@devyte, can you merge this?

**Comment by svofski** [comment]
@jandrassy sure! https://github.com/esp8266/Arduino/issues/4511

**Pull request merged**